iree/turbine/kernel/wave/templates/vanilla_attention.py

```diff
@@ -179,3 +184,279 @@ def base_attention(
+    LOAD_ELEMS_PER_THREAD_PV = index_symbol("LOAD_ELEMS_PER_THREAD_PV")
+    STORE_ELEMS_PER_THREAD = tkl.sym.STORE_ELEMS_PER_THREAD
+
+    # Expose user-constraints
+    constraints: list[tkw.Constraint] = [tkw.WorkgroupConstraint(M, BLOCK_M, 0)]
+    constraints += [tkw.WorkgroupConstraint(N, BLOCK_N, 1)]
+    constraints += [tkw.WorkgroupConstraint(B, BLOCK_B, 2)]
+    constraints += [tkw.WorkgroupConstraint(H, BLOCK_H, 3)]
+    constraints += [tkw.TilingConstraint(K2, BLOCK_K2)]
+    constraints += [tkw.WaveConstraint(M, BLOCK_M / 4)]
+    constraints += [tkw.WaveConstraint(N, BLOCK_N / 1)]
+
+    if mfma_variant[1] == MMAType.F32_16x16x16_F16:
+        Mvec = 16
+        Nvec = 16
+    if mfma_variant[1] == MMAType.F32_32x32x8_F16:
+        Mvec = 32
+        Nvec = 32
+
+    constraints += [
+        tkw.HardwareConstraint(
+            threads_per_wave=64,
+            waves_per_block=(4, 1, 1),
+            mma_type=mfma_variant[1],
+            vector_shapes={B: 0, H: 0, M: Mvec, N: Nvec},
+        )
+    ]
+
+    if dynamic_dims:
+        constraints += [tkw.Assumption(K2 > BLOCK_K2 * 4)]
+
+    i = tkw.IndexMapping.iterator(0)
+    j = tkw.IndexMapping.iterator(1)
+    k = tkw.IndexMapping.iterator(2)
+    l = tkw.IndexMapping.iterator(3)
+    mapping = tkw.IndexMapping(
+        num_iterators=4,
+        inputs={B: i, H: j, N: k, M: l},
+        outputs={B: i, M: l, H: j, N: k},
+    )
+    q_mapping = tkw.IndexMapping(
+        num_iterators=4,
+        inputs={B: i, H: j, M: k, K1: l},
+        outputs={B: i, H: j, M: k, K1: l},
+    )
+    k_mapping = tkw.IndexMapping(
+        num_iterators=4,
+        inputs={B: i, H: j, K2: k, K1: l},
+        outputs={B: i, H: j, K2: k, K1: l},
+    )
+    v_mapping = tkw.IndexMapping(
+        num_iterators=4,
+        inputs={B: i, H: j, N: k, K2: l},
+        outputs={B: i, H: j, N: k, K2: l},
+    )
+
+    log2e = 1.44269504089
+    dk_sqrt = math.sqrt(1.0 / shape.head_size)
+
+    def base_attention_core(q, k, v, c):
+        qkv_scaling = tkl.Register[B, H, M, K1, tkl.f16](dk_sqrt * log2e)
+        c_reg = tkl.Register[B, H, N, M, tkl.f32](0.0)
+        init_sum = tkl.Register[B, H, M, tkl.f32](0.0)
+        init_max = tkl.Register[B, H, M, tkl.f32](-1e6)
+        ZEROF = tkl.Register[M, K2, tkl.f32](0.0)
+        MIN_INF = tkl.Register[M, K2, tkl.f32](-1e6)
+
+        # This microkernel encodes the fact that if the reduction
+        # dimension were tiled, then we would need to materialize a loop.
+        @tkw.reduction(K2, init_args=[init_max, init_sum, c_reg])
+        def repeat(
+            partial_max: tkl.Register[B, H, M, tkl.f32],
+            partial_sum: tkl.Register[B, H, M, tkl.f32],
+            acc: tkl.Register[B, H, N, M, tkl.f32],
+        ):
+            imm_reg = tkl.Register[B, H, K2, M, tkl.f32](0.0)
+            q_reg = tkw.read(
+                q, elements_per_thread=LOAD_ELEMS_PER_THREAD_QK, mapping=q_mapping
```

Comment:
Explicit `elements_per_thread` should no longer be needed here. Can you rebase and check it?

Reply:
Addressed.